common/map_benchmark.cpp

```diff
@@ -0,0 +1,466 @@
+
+      m.BenchErase(lookup_keys[i]);
+      benchmark::ClobberMemory();
+
+      bool inserted = m.BenchUpdate(lookup_keys[i], MakeValue2<VT>());
+      CARBON_DCHECK(inserted);
+    }
+  }
+}
+MAP_BENCHMARK_ONE_OP(BM_MapEraseUpdateHit, HitArgs);
+
+// NOLINTBEGIN(bugprone-macro-parentheses): Parentheses are incorrect here.
+#define MAP_BENCHMARK_OP_SEQ_SIZE(NAME, KT, VT)                  \
+  BENCHMARK(NAME<Map<KT, VT>>)->Apply(SizeArgs);                 \
+  BENCHMARK(NAME<absl::flat_hash_map<KT, VT>>)->Apply(SizeArgs); \
+  BENCHMARK(NAME<llvm::DenseMap<KT, VT>>)->Apply(APPLY);         \
+  BENCHMARK(NAME<llvm::DenseMap<KT, VT, CarbonHashDI<KT>>>)->Apply(SizeArgs)
+// NOLINTEND(bugprone-macro-parentheses)
+
+#define MAP_BENCHMARK_OP_SEQ(NAME)                       \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, int, int);             \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, int*, int*);           \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, int, llvm::StringRef); \
+  MAP_BENCHMARK_OP_SEQ_SIZE(NAME, llvm::StringRef, int)
+
+// This is an interesting, somewhat specialized benchmark that measures the cost
+// of inserting a sequence of key/value pairs into a table with no collisions up
+// to some size and then inserting a colliding key and throwing away the table.
+//
+// This can give an idea of the cost of building up a map of a particular size,
+// but without actually using it. Or of algorithms like cycle-detection which
+// for some reason need an associative container.
+//
+// It also covers both the insert-into-an-empty-slot code path that isn't
+// covered elsewhere, and the code path for growing a table to a larger size.
+//
+// Because this benchmark operates on whole maps, we also compute the number of
+// probed keys for Carbon's set as that is both a general reflection of the
+// efficacy of the underlying hash function, and a direct factor that drives the
+// cost of these operations.
+template <typename MapT>
+static void BM_MapInsertSeq(benchmark::State& state) {
+  using MapWrapperT = MapWrapper<MapT>;
+  using KT = typename MapWrapperT::KeyT;
+  using VT = typename MapWrapperT::ValueT;
+  constexpr ssize_t LookupKeysSize = 1 << 8;
+  auto [keys, lookup_keys] =
+      GetKeysAndHitKeys<KT>(state.range(0), LookupKeysSize);
+
+  // Note that we don't force batches that use all the lookup keys because
+  // there's no difference in cache usage by covering all the different lookup
+  // keys.
+  ssize_t i = 0;
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(i);
+
+    MapWrapperT m;
+    for (auto k : keys) {
+      bool inserted = m.BenchInsert(k, MakeValue<VT>());
+      CARBON_DCHECK(inserted) << "Must be a successful insert!";
+    }
+
+    // Now insert a final random repeated key.
+    bool inserted = m.BenchInsert(lookup_keys[i], MakeValue2<VT>());
+    CARBON_DCHECK(!inserted) << "Must already be in the map!";
+
+    // Rotate through the shuffled keys.
+    i = (i + static_cast<ssize_t>(!inserted)) & (LookupKeysSize - 1);
+  }
+
```

Comment:
Shouldn't there be something telling the benchmarking infrastructure to stop timing at this point?